File: tests/java/tests-java-junit5-api_request-playwright/src/test/java/api/JsonSchemas.java
package api;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.networknt.schema.JsonSchemaFactory;
import com.networknt.schema.SpecVersion;
import com.networknt.schema.ValidationMessage;

import java.io.InputStream;
import java.util.Set;
import java.util.stream.Collectors;

import static org.junit.jupiter.api.Assertions.assertTrue;

public final class JsonSchemas {

    private static final ObjectMapper MAPPER = new ObjectMapper();
    private static final JsonSchemaFactory FACTORY = JsonSchemaFactory.getInstance(SpecVersion.VersionFlag.V7);

    private JsonSchemas() {
    }

    public static void assertMatches(String body, String name) {
        InputStream stream = JsonSchemas.class.getResourceAsStream("/schemas/" + name);
        if (stream == null) {
            throw new IllegalStateException("missing classpath schema " + name);
        }
        JsonNode node;
        try {
            node = MAPPER.readTree(body);
        } catch (Exception e) {
            throw new IllegalStateException(name + " not JSON: " + body, e);
        }
        Set<ValidationMessage> errors = FACTORY.getSchema(stream).validate(node);
        assertTrue(errors.isEmpty(), () -> name + ": "
                + errors.stream().map(ValidationMessage::getMessage).collect(Collectors.joining("; "))
                + " — " + body);
    }
}

File: tests/kotlin/tests-kotlin-junit5-api_request-playwright/src/test/kotlin/api/JsonSchemas.kt
package api

import com.fasterxml.jackson.databind.JsonNode
import com.fasterxml.jackson.databind.ObjectMapper
import com.networknt.schema.JsonSchemaFactory
import com.networknt.schema.SpecVersion
import org.junit.jupiter.api.Assertions.assertTrue

object JsonSchemas {

    private val mapper = ObjectMapper()
    private val factory = JsonSchemaFactory.getInstance(SpecVersion.VersionFlag.V7)

    fun assertMatches(body: String, name: String) {
        val stream = checkNotNull(javaClass.getResourceAsStream("/schemas/$name")) {
            "missing classpath schema $name"
        }
        val schema = factory.getSchema(stream)
        val node = mapper.readTree(body)
        val errors = schema.validate(node)
        assertTrue(errors.isEmpty()) {
            "$name: ${errors.joinToString("; ") { it.message }} — $body"
        }
    }
}

data class HttpResult(val status: Int, val body: String) {
    val json: JsonNode by lazy { MAPPER.readTree(body.ifBlank { "{}" }) }

    fun text(field: String): String = json.path(field).asText()

    fun itemNames(): List<String> =
        json.path("items").map { it.path("name").asText() }

    companion object {
        val MAPPER: ObjectMapper = ObjectMapper()
    }
}
